test/gtest/ucp/test_ucp_proto_mock.cc

```diff
@@ -34,71 +34,117 @@ class mock_iface {
+
         ucs_list_for_each(component, &uct_components_list, list) {
             uct_tl_t *tl;
             ucs_list_for_each(tl, &component->tl_list, list) {
                 if (tl_name == tl->name) {
-                    EXPECT_EQ(0, m_tls.count(tl_name));
-                    m_tls[tl_name] = tl;
+                    m_mock.setup(&component->query_md_resources, query_md_mock);
+                    m_mock.setup(&tl->query_devices, query_devices_mock);
                     m_mock.setup(&tl->iface_open, iface_open_mock);
+                    m_tl = tl;
+                    return;
                 }
             }
         }
+
+        FAIL() << "Transport " << tl_name << " not found";
+    }
+
+private:
+    static ucs_status_t query_md_mock(uct_component_t *component,
+                                      uct_md_resource_desc_t **resources_p,
+                                      unsigned *num_resources_p)
+    {
+        UCS_MOCK_ORIG_FUNC(m_self->m_mock, &component->query_md_resources,
+                           component, resources_p, num_resources_p);
+        /* Keep only the first available MD */
+        *num_resources_p = ucs_min(*num_resources_p, 1);
+        return UCS_OK;
+    }
+
+    static ucs_status_t
+    query_devices_mock(uct_md_h md, uct_tl_device_resource_t **tl_devices_p,
+                       unsigned *num_tl_devices_p)
+    {
+        UCS_MOCK_ORIG_FUNC(m_self->m_mock, &m_self->m_tl->query_devices, md,
+                           tl_devices_p, num_tl_devices_p);
+        if (*num_tl_devices_p == 0) {
+            return UCS_OK;
+        }
+
+        /*
+         * Store first device name, that will be used to instantiate all the
+         * mocks. Then mock device query response to return mocks instead.
+         */
+        m_self->m_real_dev_name  = (*tl_devices_p)[0].name;
+        unsigned size            = m_self->m_iface_attrs_funcs.size();
+        auto tmp                 = (uct_tl_device_resource_t *)ucs_malloc(
+                                        sizeof(uct_tl_device_resource_t) * size,
+                                        "mock_tl_devices");
+        ucs_sys_device_t sys_dev = 1;
+        unsigned i               = 0;
+        for (const auto &it : m_self->m_iface_attrs_funcs) {
+            ucs_strncpy_safe(tmp[i].name, it.first.c_str(), UCT_DEVICE_NAME_MAX);
+            tmp[i].type         = (*tl_devices_p)[0].type;
+            tmp[i++].sys_device = sys_dev++;
+        }
+
+        ucs_free(*tl_devices_p);
+        *tl_devices_p     = tmp;
+        *num_tl_devices_p = size;
+        return UCS_OK;
     }
 
     static ucs_status_t
     iface_open_mock(uct_md_h md, uct_worker_h worker,
                     const uct_iface_params_t *params,
                     const uct_iface_config_t *config, uct_iface_h *iface_p)
     {
-        uct_tl_t *tl = m_self->m_tls[params->mode.device.tl_name];
-        ucs_status_t status;
+        uct_iface_params_t copy_params   = *params;
```

Comment:
1. copy_params -> iface_open_params
2. do we need to check presence of UCT_IFACE_PARAM_FIELD_DEVICE? assert?